lib/Pipeline/ExecutionContext.cpp

```diff
@@ -0,0 +1,40 @@
+/// \file ExecutionContext::RAIICommitter.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+//
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/ExecutionContext.h"
+#include "revng/Pipeline/Step.h"
+#include "revng/Pipeline/Target.h"
+
+using namespace pipeline;
+
+ExecutionContext::ExecutionContext(Context &Ctx,
+                                   Step &Step,
+                                   PipeWrapper *Pipe) :
+  Ctx(&Ctx), CurrentStep(&Step), Pipe(Pipe) {
+  if (Pipe != nullptr)
+    getContext().clearReadFields();
+}
+
+pipeline::ExecutionContext::~ExecutionContext() {
+  if (Pipe != nullptr)
+    getContext().stopTracking();
+}
+
+void ExecutionContext::commit(const Target &Target,
+                              llvm::StringRef ContainerName) {
+  TargetInStep ToCollect(Target, ContainerName.str());
+  getContext().collectReadFields(ToCollect,
+                                 Pipe->ContainerWiseCache.getPathCache());
+
+  getContext().clearReadFields();
```

Comment:
i can, but it effectively eats away from the maximum number of possible elements in the stack. Not sure we care about that.

Reply:
> it effectively eats away from the maximum number of possible elements in the stack

Whta is "it"? push + pop eats only on element on the stack. `clearReadFields` acts only on the topmost one AFAIU.

Alternatively, we can rename `clearReadFields`, but I'm not sure it's easy to come up with a nice name saying that only the topmost layer is dropped.

Reply:
i considered to implement it with pop followed by push, but that is broken when the stack was empty, since you can't pop a empty stack. i could place a check and not do the pop if it is empty, but i would prefer to keep that as an assertion. i guess clearReadFields works as a name

Reply:
i went for clearTop